_posts/2020-06-05-cruise-control.md

```diff
@@ -0,0 +1,213 @@
+
+<!--more-->
+
+## Cruise Control overview
+
+Cruise Control started out as an internal project at Linkedin and was used for managing their in-house Kafka clusters. It was open sourced in 2016 and has seen constant development since then. 
+
+Cruise Control consists of a main server implementation that contains several sub-systems for monitoring, analyzing and proposing changes to the Kafka Cluster it monitors. As well as the main server, Cruise Control provides a Kafka metrics reporter implementation that, once installed into the Kafka brokers, records a wide range of metrics in addition to the default metrics provided by the brokers themselves.
+
+The Cruise Control server periodically collects the metrics recorded by the metrics reporters (which are sent to topics within the cluster) to infer the traffic pattern of each partition. Based on the traffic characteristics and distribution of all the partitions, it derives the load impact of each partition on the brokers. Cruise Control then builds a workload model to simulate the workload of the Kafka cluster. 
+
+The user can define a list of _optimization goals_. These goals each define a set of parameter values that the cluster state should meet. For example a goal can enforce that network traffic for each broker should not differ by more than +/- 10% of the cluster's average. These goals are configured in a priority order and when required the Cruise Control server's analyzer sub-system will attempt to change the workload model to meet each goal. These changes consist of switching the lead replica of a given partition replica leadership or reassigning partitions to other brokers. The analyzer will descend down the list of goals and first check if the current cluster configuration meets that goal and if it doesn't will attempt to alter the cluster model until it does. At each step any generated changes are passed to all the previous goals in the list to check if they will then violate those higher priority goals.
+
+Each of the goals can either be _hard_ or _soft_. A hard goal **must** be satisfied, if the analyzer proposes changes that violate a hard goal then those changes are rejected. A soft goal can be violated. Therefore an _optimization proposal_, which is a collection of changes resulting from a full pass over a given list of goals, will meet the requirements of all the hard goals in that list but may not satisfy all the soft goals. It is worth noting that if it is not possible to meet all the hard goals then no optimization proposal will be produced.
```

Comment:
```all of the soft goal```

Not completely sure if the "of" is needed here but my brain inserts it when I read this line, do what you will!